File: summary/ruby/13_bfs.rb
def breadth_first_search(graph, start_vertex)
  queue = [start_vertex]
  visited = {start_vertex => true}
  order_of_visit = []

  while queue.any?
    current_vertex = queue.shift
    order_of_visit << current_vertex

    graph[current_vertex].each do |neighbor|
      unless visited[neighbor]
        queue.push(neighbor)
        visited[neighbor] = true
      end
    end
  end

  order_of_visit
end

# Example usage:
graph = {
  'A' => ['B', 'C'],
  'B' => ['A', 'D', 'E'],
  'C' => ['A', 'F'],
  'D' => ['B'],
  'E' => ['B', 'F'],
  'F' => ['C', 'E']
}
puts breadth_first_search(graph, 'A').inspect

File: summary/ts/13_bfs.ts
function breadthFirstSearch(graph: Record<string, string[]>, startVertex: string): string[] {
    const queue: string[] = [startVertex];
    const visited: Record<string, boolean> = { [startVertex]: true };
    const orderOfVisit: string[] = [];

    while (queue.length > 0) {
        const currentVertex = queue.shift()!;
        orderOfVisit.push(currentVertex);

        graph[currentVertex].forEach(neighbor => {
            if (!visited[neighbor]) {
                queue.push(neighbor);
                visited[neighbor] = true;
            }
        });
    }

    return orderOfVisit;
}

// Example usage:
const graph: Record<string, string[]> = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'E']
};
console.log(breadthFirstSearch(graph, 'A'));
